Support trace API similar to OpenEthereum/Parity's one

ziogaschr opened the pull request:
This PR adds a compatible trace API with [OpenEthereum's one](https://openethereum.github.io/wiki/JSONRPC-trace-module#trace_block). 

At the moment we have implemented:

- [x] trace_transaction
- [x] trace_block

For a node to enable this feature, it has to start with `--vmdebug` flag.

For the best performance node has to be synced in **archive** mode.
In console the `trace` API can be used.
Last, the `trace` RPC API is available, though it have to be enabled.

Fixes  [#100](https://github.com/etclabscore/core-geth/issues/100).

---

**Work in Progress. Opened a PR to handle discussion over here.**

ziogaschr committed:
tracer: Add call_tracer_open_ethereum with formatted output similar to OpenEthereum

# Conflicts:
#	eth/tracers/internal/tracers/assets.go

ziogaschr committed:
tracer: Expose trace_traceTransaction API (draft, might change)

ziogaschr committed:
tracer: Add `trace_traceBlock*` methods

ziogaschr committed:
Rename TraceBlockByNumber to TraceBlock for compatibility with OpenEthereum

ziogaschr committed:
Rename trace methods for compatibility with OpenEthereum names

ziogaschr committed:
tracer: add dump trace_filter func

ziogaschr committed:
tracer: include `blockHash`, `transactionHash` and `transactionPosition` in response

ziogaschr committed:
tracer: small change to accumulateRewards to make it available to the RPC

Based on @tjayrush commit https://github.com/ledgerwatch/turbo-geth/commit/47f59e8fcd805ad2a4f677f7dffb2ae9b0673bbc

ziogaschr committed:
tracer: restructure PrivateTraceAPI to its own file

ziogaschr committed:
tracer: add block/uncle rewards in trace_block

ziogaschr committed:
tracer: reorder tracer json fields

ziogaschr committed:
Cleanup

ziogaschr committed:
tracer: handle gas 2300 stipend

ziogaschr committed:
tracer: Add mapping of core-geth errors to OpenEthereum ones

ziogaschr committed:
tracer: handle op=selfdestruct similar to OpenEthereum

ziogaschr committed:
tracer: cleanup reward structs

ziogaschr committed:
tracer: JS better handling of typeof undefined

ziogaschr committed:
tracer: JS rename oeErrorMapping

ziogaschr committed:
tracer: JS cleanup 🧽

ziogaschr committed:
tracer: JS refactor/cleanup finalize()

ziogaschr committed:
tracer: JS add additional error maps

ziogaschr committed:
js/tracers: make OpenEthereum calltracer report value in selfdestructs

js/tracers: make calltracer report value in selfdestructs (#21549)

https://github.com/ethereum/go-ethereum/commit/71c37d82adaa2b69ea98ce0c5505489d6b711c1e

ziogaschr committed:
tracer: JS better handling of gas and gasUsed (remove the stipend hack)

ziogaschr committed:
tracer: JS OE compatibility fixes

ziogaschr committed:
tracer: JS handle out of gas errors

ziogaschr committed:
tracer: JS temp hack for compatibility (to be changed)

ziogaschr committed:
tracer: fix handling of op=delegatecall value

.

ziogaschr committed:
tracer: better naming for isDynamicGas variable (still not sure if accurate)

ziogaschr committed:
tracer: add more error mappings

ziogaschr committed:
tracer: op=CREATE2 to return type=CREATE for OE compatibility

ziogaschr commented on eth/tracers/tracer.go:
```diff
@@ -541,6 +541,13 @@ func (jst *Tracer) CaptureStart(from common.Address, to common.Address, create b
 	return nil
 }
 
+func (jst *Tracer) CaptureExtraContext(inputs map[string]interface{}) error {
```
@iquidus @meowsbits I am thinking to make the `map[string]interface{}` a Struct and parse it in here with reflect. Do you think it's ok for now like it is? Is it overkill to use reflect for this?

meowsbits replied:
For now I think it would be overkill... although worth considering for a refactor later on. The `Tracer.ctx` field is used in several ways that seem to depend on the type-agnosticism of the `map[string]interface{}`, eg. `txTraceTask.extraContext`, so this may be something of a rabbit hole.

ziogaschr replied:
That's true, originally `Tracer.ctx` holds transaction context. I have added `CaptureExtraContext` in order to be able to extend context with block related stuff. For this reason I was thinking to make only the `inputs` in `CaptureExtraContext` a struct. Though at the end, they will be added down to the `Tracer.ctx` being a `map[string]interface{}`

meowsbits committed:
eth: clean ineffectual assigns re: OE tracer config setting

The function setConfigTracerToOpenEthereum mutates
the provided config value by pointer.

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
eth: (lint) remove unnecessary type conversion

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
eth: (lint:gosimple) replace iterating assign with variadic append

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
main: add 'trace:1.0' to console test

This module is added, and the test fails because it
has not been update to include the addition to expected
IPC APIs.
This fixes that.

Signed-off-by: meows <b5c6@protonmail.com>

ziogaschr commented on eth/api_tracer_oe.go:
```diff
@@ -155,9 +155,7 @@ func (api *PrivateTraceAPI) Block(ctx context.Context, number rpc.BlockNumber, c
-		for _, item := range tmp {
-			results = append(results, item)
-		}
+		results = append(results, tmp...)
```
I haven't run actual perf tests, but I've read somewhere that this might be much slower on bigger lists (which is not the case here).

meowsbits replied:
Interesting! I just wrote some quick benchmarks to investigate, included in https://github.com/etclabscore/core-geth/pull/211/commits/69d7e09ae8b6d41889de30d328b48ffdecbe4469.

```go
// BenchmarkTraceResultsAppend1 compares performance against BenchmarkTraceResultsAppend2,
// comparing the performance of different ways of appending items to slices.
// This is used in PrivateTraceAPI#Block appending results to the traceResults value.
func BenchmarkTraceResultsAppend1(b *testing.B) {
	traceResults := []interface{}{}
	for i := 0; i < 10000; i++ {
		traceResults = append(traceResults, i)
	}

	b.ResetTimer()

	for i := 0; i < b.N; i++ {
		// results := make([]interface{}, len(traceResults))
		results := []interface{}{}
		for _, it := range traceResults {
			results = append(results, it)
		}
	}
}

// BenchmarkTraceResultsAppend2 (see comment for BenchmarkTraceResultsAppend1).
func BenchmarkTraceResultsAppend2(b *testing.B) {
	traceResults := []interface{}{}
	for i := 0; i < 10000; i++ {
		traceResults = append(traceResults, i)
	}

	b.ResetTimer()

	for i := 0; i < b.N; i++ {
		// results := make([]interface{}, len(traceResults))
		results := []interface{}{}
		results = append(results, traceResults...)
	}
}
```

Here are the results:

```txt
goos: linux
goarch: amd64
pkg: github.com/ethereum/go-ethereum/eth
BenchmarkTraceResultsAppend1
BenchmarkTraceResultsAppend1-12    	    3742	    288702 ns/op
BenchmarkTraceResultsAppend2
BenchmarkTraceResultsAppend2-12    	   20686	     59821 ns/op
PASS
```

ziogaschr replied:
I love you man. Code says always the truth.

meowsbits committed:
eth: add benchmarking test for slice append patterns

Rel to an interesting comment
https://github.com/etclabscore/core-geth/pull/211#pullrequestreview-509441839

This is likely NOT a very important performance
comparison, since the result values are of
limited length and tracing is expected to be asyncronous
to client/chain import. But interesting, anyways.

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits commented on eth/api_tracer.go:
```diff
@@ -352,23 +353,29 @@ func (api *PrivateDebugAPI) traceChain(ctx context.Context, start, end *types.Bl
 // TraceBlockByNumber returns the structured logs created during the execution of
 // EVM and returns them as a JSON object.
-func (api *PrivateDebugAPI) TraceBlockByNumber(ctx context.Context, number rpc.BlockNumber, config *TraceConfig) ([]*txTraceResult, error) {
+func traceBlockByNumber(ctx context.Context, eth *Ethereum, number rpc.BlockNumber, config *TraceConfig) ([]*txTraceResult, error) {
```
What's the reasoning behind pulling the receiver out and using the `eth *Ethereum` arg instead? (Instead of just making the method private?) Not against it, just curious.

ziogaschr replied:
Just for using it in both **PrivateDebugAPI** and **PrivateTraceAPI** [here](https://github.com/etclabscore/core-geth/pull/211/files#diff-e98ceee1b6a777b2ee45a6f6527b9d51da990c691cbb19997c8c104ca9dd4725R136). I would be interested if you have a better idea on doing this, as I don't like it much. I did the same for more functions.

meowsbits commented on eth/api_tracer_oe.go:
```diff
@@ -0,0 +1,181 @@
+// OpenEthereumTrace A trace in the desired format (Parity/OpenEtherum) See: https://openethereum.github.io/wiki/JSONRPC-trace-module
+type OpenEthereumTrace struct {
+	Action              TraceRewardAction `json:"action"`
+	BlockHash           *common.Hash      `json:"blockHash"`
```
Does this need to be a pointer? Since `BlockNumber` isn't, and neither type uses `omitempty`, it seems there is an assumption that the block context will always be available. Or?

ziogaschr replied:
No it doesn't. Nice catch. It actually started wrong from wrong assumptions from OpenEthereum trace outputs and copying them from turbo-geth as starting structs.

ziogaschr replied:
@meowsbits if we want to keep 100% compatibility with OpenEthereum tracer, then they return the below for rewards, as they use the same struct for both `trace_block` and `trace_transaction`.

```
    {
            "action": {
                "author": "0x877bd459c9b7d8576b44e59e09d076c25946f443",
                "rewardType": "block",
                "value": "0x4563918244f40000"
            },
            "blockHash": "0x8df024322173d225a09681d35edeaa528aca60743a11a70f854c158862bf5282",
            "blockNumber": 553416,
            "result": null,
            "subtraces": 0,
            "traceAddress": [],
            "transactionHash": null,
            "transactionPosition": null,
            "type": "reward"
        }
```

In our case where we use it only for block rewards, it's totally fine to remove the `TransactionHash` and `TransactionPosition`. What do you think? Shall we keep them?

meowsbits commented on eth/api_tracer_oe.go:
```diff
@@ -0,0 +1,181 @@
+	case rpc.PendingBlockNumber:
+		block = api.eth.miner.PendingBlock()
+	case rpc.LatestBlockNumber:
+		block = api.eth.blockchain.CurrentBlock()
```
Should we handle `case rpc.EarliestBlockNumber` here too? 

EDIT: Actually the method and apparently some of associated "helper" functions don't support genesis block tracing. Which makes sense, I guess, but seems kind of silently assumed... ?

ziogaschr replied:
Probably not, I copied this from the `debug_traceBlock` where it's neither being handled. The reason it's not being handled is, that we can't retrieve block trace for the genesis block. We can add it if a reason exists, though the current logic fails when it tries to read the state of the parent block, as genesis has no parent block.

p.s.: I have extracted the logic of getting a block from a block number in a stash I have for `trace_filter` support.

ziogaschr commented on eth/api_tracer_oe.go:
```diff
@@ -50,7 +50,7 @@ type TraceRewardAction struct {
-// setConfigTracerToOpenEthereum forces the Tracer to the OpenEthereum one
+// setConfigTracerToOpenEthereum forces the Tracer to the OpenEthereum one, mutating the argument value.
 func setConfigTracerToOpenEthereum(config *TraceConfig) *TraceConfig {
 	if config == nil {
```
@meowsbits when `config == nil`, there is no actual pointer to assign to. In this case the config is being scoped in `setConfigTracerToOpenEthereum`. Assigning to an indirect pointer, can't work in this case either. If we want to do it without assigning from the caller func, then we have to call it like:

```go
if config == nil {
	config = new(TraceConfig)
}
setConfigTracerToOpenEthereum(config)
```

ziogaschr replied:
@meowsbits what do you think on this? do you have any better ideas?

ziogaschr replied:
Reverted for now, so as it works on external review. Will leave this review unresolved though.

ziogaschr committed:
tracer: remove unused OpenEthereumTrace fields

We use OpenEthereumTrace only to trace block rewards for now

ziogaschr committed:
tracer: JS cleanup

ziogaschr committed:
tracer: Set value=0x0 for op=STATICCALL

ziogaschr committed:
tracer: JS handle output based on op=RETURN

ziogaschr committed:
tracer: JS remove the uneeded isDynamic check on gas

ziogaschr committed:
tracer: JS on “out of gas” Failure set gas=0x0

ziogaschr committed:
tracer: JS fix

ziogaschr committed:
tracer: revert remove unused OpenEthereumTrace fields

ziogaschr committed:
tracer: JS OE handle “return data out of bounds”

ziogaschr committed:
tracer: JS OE on fault, try to get gas from inner calls

ziogaschr committed:
Revert "eth: clean ineffectual assigns re: OE tracer config setting"

This reverts commit 6099dbc86bfa4e81516baa7008a2839e3f9ef01c.

ziogaschr committed:
core/vm: track 63/64 call gas off stack

Apply logic from commit hash be12392fbad9f4a861130a347e6bcf07a5c34974 to CREATE and CREATE2 opcodes.

Gas calculations in gasCall* relayed the available gas for calls by
replacing it on the stack. This lead to inconsistent traces, which we
papered over by copying the pre-execution stack in trace mode.

This change relays available gas using a temporary variable, off the
stack, and allows removing the weird copy.

ziogaschr committed:
tracer: pass the true available gas for an opcode call in CaptureState

ziogaschr committed:
tracer/js: extend JS log tracer with getAvailableGas per call

ziogaschr committed:
tracer/js: extend JS log tracer with getReturnData

ziogaschr committed:
tracer/js: add "Out of stack” error mapping for OE tracer

ziogaschr committed:
tracer/js: handle edje cases where first time in callstack remains empty while next one has data

ziogaschr committed:
tracer/js: extend OE JS tracer to return compatible gas with OE using getAvailableGas

ziogaschr committed:
tracer/js: extend OE JS tracer to use getReturnData to decide if the call returned any data

ziogaschr committed:
tracer/js: bring back wrongly removed line

ziogaschr committed:
cleanup

ziogaschr committed:
tracer/js: handle precompiled contract failures on CALL opcodes

ziogaschr committed:
tracer/js: expose system opcode errors in JS tracers

This commit allows the system opcode errors to be read from within the JS tracers. It does that in a very hacky way for now just to allow OpenEthereum tracer 100% compatibility.

After some discussion on better solutions, we might add a new PR with a better way of handling those.

ziogaschr committed:
cleanup

ziogaschr committed:
tracer/js: read system opcode error in OE JS tracer

ziogaschr committed:
tracer/js: extend OE JS tracer to use only getReturnData for CALL related opcodes output

ziogaschr committed:
tracers: generate assets.go

ziogaschr committed:
tracers/js: cleanup

ziogaschr committed:
tracer: Rename call_tracer_open_ethereum to call_tracer_parity and rest references

ziogaschr committed:
tracer: fix lint issues

ziogaschr committed:
tracer/js: add TODO comment

ziogaschr committed:
core/vm: track 63/64 create gas off stack (fix)

ziogaschr committed:
tracers: generate assets.go

ziogaschr committed:
core/vm: add createGasEip150 for CREATE/CREATE2 gas calc of 63/64 rule on EIP150

ziogaschr committed:
core/vm: cleanup use of evm.callGasTemp in interpreter

ziogaschr committed:
Merge branch 'master' into feat/openethereum-trace-apis

ziogaschr commented on consensus/ethash/consensus.go:
```diff
@@ -587,28 +587,43 @@ var (
 // reward. The total reward consists of the static block reward and rewards for
 // included uncles. The coinbase of each uncle block is also rewarded.
-func accumulateRewards(config ctypes.ChainConfigurator, state *state.StateDB, header *types.Header, uncles []*types.Header) {
+// func AccumulateRewards(config ctypes.ChainConfigurator, state *state.StateDB, header *types.Header, uncles []*types.Header) {
```
@meowsbits @iquidus can you review this function and the ones that being used inside for any logical issues?

meowsbits replied:
Have done. LGTM, and the passing tests indicate positive too.

iquidus commented on consensus/ethash/consensus.go:
```diff
@@ -587,28 +587,43 @@ var (
 // AccumulateRewards credits the coinbase of the given block with the mining
```
This description no longer matches the function.

iquidus commented on consensus/ethash/consensus.go:
```diff
@@ -587,28 +587,43 @@ var (
 // included uncles. The coinbase of each uncle block is also rewarded.
-func accumulateRewards(config ctypes.ChainConfigurator, state *state.StateDB, header *types.Header, uncles []*types.Header) {
+// func AccumulateRewards(config ctypes.ChainConfigurator, state *state.StateDB, header *types.Header, uncles []*types.Header) {
+func AccumulateRewards(config ctypes.ChainConfigurator, header *types.Header, uncles []*types.Header) (*big.Int, []*big.Int) {
```
Perhaps rename func to `CalcRewards`? Or for consistency with reward calculation func's `GetRewards`

meowsbits commented on core/vm/logger.go:
```diff
@@ -106,7 +106,7 @@ func (s *StructLog) ErrorString() string {
 type Tracer interface {
 	CaptureStart(from common.Address, to common.Address, create bool, input []byte, gas uint64, value *big.Int) error
-	CaptureState(env *EVM, pc uint64, op OpCode, gas, cost uint64, memory *Memory, stack *Stack, rStack *ReturnStack, rData []byte, contract *Contract, depth int, err error) error
+	CaptureState(env *EVM, pc uint64, op OpCode, gas, availableGas, cost uint64, memory *Memory, stack *Stack, rStack *ReturnStack, rData []byte, contract *Contract, depth int, err error) error
```
@ziogaschr I think you mentioned yesterday that this parameter may be removable since this value is equivalent to `env.callGasTemp`. If so, is it worth doing?

ziogaschr replied:
At this point removed the extra `availableGasCopy` and use the `in.evm.callGasTemp`. This one can be removed if I make it public as `in.evm.CallGasTemp`. What do you think about it?

meowsbits replied:
I think it's OK to make it public because that follows the pattern that we've already "allowed" with the introduction of `EVM.CallErrorTemp`, and IMO publicizing a field there is less "expensive" than adding a parameter to a shared interface.

ziogaschr committed:
consensus/ethash: better name for AccumulateRewards -> GetRewards

based on @iquidus suggestion

meowsbits commented on consensus/ethash/consensus.go:
```diff
@@ -584,31 +584,46 @@ var (
+func accumulateRewards(config ctypes.ChainConfigurator, state *state.StateDB, header *types.Header, uncles []*types.Header) {
+	minerReward, uncleRewards := GetRewards(config, header, uncles)
+	for i, uncle := range uncles {
+		if i < len(uncleRewards) {
```
This is an unnecessary condition since each uncle will always get one reward (and they're in a known and constant order), so `len(uncleRewards)` will always be `len(uncles)`.

meowsbits committed:
ethash: remove unnecessarily defensive conditional

This is an unnecessary condition since each uncle will always get one reward (and they're in a known and constant order), so len(uncleRewards) will always be len(uncles).

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits reviewed (dismissed):
:x: `go test ./eth/tracers/...` is failing

Eg. https://travis-ci.org/github/etclabscore/core-geth/jobs/741433701#L960

ziogaschr committed:
core/vm, eth/tracers: make `evm.callGasTemp` public `evm.CallGasTemp`

ziogaschr committed:
eth/tracers: fix broken test `call_tracer_inner_create_oog_outer_throw.json`
